### ISSUE TITLE
[RLlib] Possible bug in TorchSquashedGaussian plus associated feature request

### ISSUE THREAD

**Issue by TristanKalloniatis** [ray-project/ray#51544]
### What happened + What you expected to happen

I am looking at the TorchSquashedGaussian distribution and have a couple of questions:

1: I don't understand why the deterministic distribution is defined as TorchDeterministic(loc=self.loc) rather than TorchDeterministic(loc=self._squash(self.loc)) - this causes the deterministic samples to not be the MLE (compare with the tf SquashedGaussian, which does this:

    def deterministic_sample(self) -> TensorType:
        mean = self.distr.mean()
        return self._squash(mean)

)

2: Why not define the KL at least, since it would just be equal to the untransformed KL (see eg Wikipedia article on KL: "Relative entropy remains well-defined for continuous distributions, and furthermore is invariant under [parameter transformations](https://en.wikipedia.org/wiki/Change_of_variables).")? I understand the entropy is more complicated since it depends on a sample, though for what it's worth other libraries do this (eg brax: https://github.com/google/brax/blob/main/brax/training/distribution.py#L83). For this question, same applies to the tf version.

Concretely I am suggesting to replace the torch version with something like this:

`
@DeveloperAPI
class TorchSquashedGaussian(TorchDistribution):
    @override(TorchDistribution)
    def __init__(
        self,
        loc: Union[float, "torch.Tensor"],
        scale: Optional[Union[float, "torch.Tensor"]] = 1.0,
        low: float = -1.0,
        high: float = 1.0,
    ):
        self.loc = loc
        self.low = low
        self.high = high

        super().__init__(loc=loc, scale=scale)

    def _get_torch_distribution(self, loc, scale) -> "torch.distributions.Distribution":
        return torch.distributions.normal.Normal(loc, scale, validate_args=False)

    @override(TorchDistribution)
    def sample(
        self, *, sample_shape=None
    ) -> Union[TensorType, Tuple[TensorType, TensorType]]:
        # Sample from the Normal distribution.
        sample = super().sample(
            sample_shape=sample_shape if sample_shape is not None else torch.Size()
        )
        # Return the squashed sample.
        return self._squash(sample)

    @override(TorchDistribution)
    def rsample(
        self, *, sample_shape=None
    ) -> Union[TensorType, Tuple[TensorType, TensorType]]:
        # Sample from the Normal distribution.
        sample = super().rsample(
            sample_shape=sample_shape if sample_shape is not None else torch.Size()
        )
        # Return the squashed sample.
        return self._squash(sample)

    @override(TorchDistribution)
    def logp(self, value: TensorType, **kwargs) -> TensorType:
        # Unsquash value.
        value = self._unsquash(value)
        # Get log-probabilities from Normal distribution.
        logp = super().logp(value, **kwargs)
        # Clip the log probabilities as a safeguard and sum.
        logp = torch.clamp(logp, -100, 100).sum(-1)
        # Return the log probabilities for squashed Normal.
        value = torch.tanh(value)
        return logp - torch.log(1 - value**2 + SMALL_NUMBER).sum(-1)

    @override(TorchDistribution)
    def entropy(self) -> TensorType:
        untransformed_entropy = super().entropy().sum(-1)
        x = self._dist.sample()
        return untransformed_entropy + torch.log(
            1 - self._squash(x) ** 2 + SMALL_NUMBER
        ).sum(-1)

    @override(TorchDistribution)
    def kl(self, other: Distribution) -> TensorType:
        return super().kl(other).sum(-1)

    def _squash(self, sample: TensorType) -> TensorType:
        # Rescale the sample to interval given by the bounds (including the bounds).
        sample = ((torch.tanh(sample) + 1.0) / 2.0) * (self.high - self.low) + self.low
        # Return a clipped sample to comply with the bounds.
        return torch.clamp(sample, self.low, self.high)

    def _unsquash(self, sample: TensorType) -> TensorType:
        # Rescale to [-1.0, 1.0].
        sample = (sample - self.low) / (self.high - self.low) * 2.0 - 1.0
        # Stabilize input to atanh function.
        sample = torch.clamp(sample, -1.0 + SMALL_NUMBER, 1.0 - SMALL_NUMBER)
        return torch.atanh(sample)

    @staticmethod
    @override(Distribution)
    def required_input_dim(space: gym.Space, **kwargs) -> int:
        assert isinstance(space, gym.spaces.Box), space
        return int(np.prod(space.shape, dtype=np.int32) * 2)

    @classmethod
    @override(TorchDistribution)
    def from_logits(
        cls, logits: TensorType, low: float = -1.0, high: float = 1.0, **kwargs
    ) -> "TorchSquashedGaussian":
        loc, log_std = logits.chunk(2, dim=-1)
        # Clip the `scale` values (coming from the `RLModule.forward()`) to
        # reasonable values.
        log_std = torch.clamp(log_std, MIN_LOG_NN_OUTPUT, MAX_LOG_NN_OUTPUT)
        scale = log_std.exp()

        # Assert that `low` is smaller than `high`.
        assert np.all(np.less(low, high))
        # Return class instance.
        return TorchSquashedGaussian(loc=loc, scale=scale, low=low, high=high)

    def to_deterministic(self) -> Distribution:
        return TorchDeterministic(loc=self._squash(self.loc))
`

### Versions / Dependencies

ray==2.43.0
Python 3.10.16
MacOS

(but I don't think any of this matters for this purpose)

### Reproduction script

`
import torch
from ray.rllib.models.torch.torch_distributions import TorchSquashedGaussian
dist = TorchSquashedGaussian(torch.tensor([1.]), torch.tensor([1.]))
deterministic_dist = dist.to_deterministic()
deterministic_dist.sample()
tensor([1.])
`

### Issue Severity

None
